mqtt/mqtt-broker/src/edgehub_auth.rs

```diff
@@ -0,0 +1,370 @@
+impl Authenticator for EdgeHubAuthenticator {
+    async fn authenticate(
+        &self,
+        username: Option<String>,
+        credentials: Credentials,
+    ) -> Result<Option<AuthId>, Error> {
+        self.internal_authenticate((username, credentials).into())
+            .await
+            .context(ErrorKind::AuthExecution)?
+            .into_auth_id()
+    }
+}
+
+impl EdgeHubAuthenticator {
+    async fn internal_authenticate(
+        &self,
+        request: EdgeHubAuthRequest,
+    ) -> Result<EdgeHubAuthResponse, Error> {
+        let response = reqwest::Client::new()
+            .post(&self.url)
+            .json(&request)
+            .send()
+            .await
+            .context(ErrorKind::AuthSendRequest)?;
+
+        if response.status() != reqwest::StatusCode::OK {
+            return Err(ErrorKind::AuthProcessResponse.into());
+        }
+
+        let response = response
+            .json::<EdgeHubAuthResponse>()
+            .await
+            .context(ErrorKind::AuthProcessResponse)?;
+
+        if !response.version.eq(API_VERSION) {
+            return Err(ErrorKind::AuthApiVersion(response.version).into());
+        }
+
+        Ok(response)
+    }
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "snake_case")]
+pub struct EdgeHubAuthRequest {
+    version: String,
+    username: Option<String>,
+    password: Option<String>,
+    certificate: Option<String>,
+    certificate_chain: Option<Vec<String>>,
+}
+
+impl From<(Option<String>, Credentials)> for EdgeHubAuthRequest {
+    fn from(item: (Option<String>, Credentials)) -> Self {
+        let (password, certificate) = match item.1 {
+            Credentials::Password(p) => (p, None),
+            Credentials::ClientCertificate(c) => (None, Some(encode(c))),
+        };
+
+        EdgeHubAuthRequest {
+            version: API_VERSION.to_string(),
+            username: item.0,
+            password,
+            certificate,
+            certificate_chain: None,
+        }
+    }
+}
+
+#[derive(Copy, Clone)]
+#[repr(i32)]
+pub enum EdgeHubResultCode {
+    Authenticated = 200,
+    Unauthenticated = 403,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct EdgeHubAuthResponse {
+    version: String,
+    result: EdgeHubResultCode,
+    identity: Option<String>,
+}
+
+impl EdgeHubAuthResponse {
+    fn into_auth_id(self) -> Result<Option<AuthId>, Error> {
+        match self.result {
+            EdgeHubResultCode::Authenticated => self
+                .identity
+                .map_or(Err(ErrorKind::AuthProcessResponse.into()), |i| {
+                    Ok(Some(AuthId::Value(i)))
+                }),
+            EdgeHubResultCode::Unauthenticated => Ok(None),
+        }
+    }
+}
+
+macro_rules! impl_for{
+    ($fun_name:ident, $typ:ident) => {
+        fn $fun_name<E: serde::de::Error>(self, v: $typ) -> Result<EdgeHubResultCode, E> {
+            match v {
+                200 => Ok(EdgeHubResultCode::Authenticated),
```

Comment:
Is there a way to avoid double defining this relation? The enum already defines Authenticated as 200 and Unauthenticated as 403, can you use that?